### ISSUE TITLE
Loss of colors at JPEG images....?!

### ISSUE THREAD

**Issue by flobox** [mosbth/cimage#120]
Hello Mikael,

I just experienced, that cimage produces a loss of colors (for JPEG images).

Please see attached images.

ORIGINAL:
![original](https://cloud.githubusercontent.com/assets/3763232/10601525/aea082e0-7712-11e5-84c2-e8c246546936.jpg)

CIMAGE:
![cimage](https://cloud.githubusercontent.com/assets/3763232/10601523/ae9b8934-7712-11e5-8781-39c4e14775ab.jpg)

imageresizing.net:
![imageresizing-asp-net](https://cloud.githubusercontent.com/assets/3763232/10601524/ae9d7424-7712-11e5-8cf2-6873175854dd.jpg)

I was using this operator: "?w=800"

Anything I can do by the config file? 

Thank you in advance!

Best,
Florian

**Comment by mosbth** [comment]
Seems like the CImage-image is half the size compared to imageresizing-version. 

```
desktop:~/git/cimage/webroot/img/issue120$ identify *                           
cimage.jpg JPEG 800x533 800x533+0+0 8-bit sRGB 44.9KB 0.000u 0:00.000           
imageresizing.jpg[1] JPEG 800x533 800x533+0+0 8-bit sRGB 100KB 0.000u 0:00.000  
orig.jpg[2] JPEG 3700x2467 3700x2467+0+0 8-bit sRGB 8.915MB 0.000u 0:00.000     
```

```
desktop:~/git/cimage/webroot/img/issue120$ file *                                                                                       
cimage.jpg:        JPEG image data, JFIF standard 1.01, resolution (DPI), density 96x96, segment length 16, comment: "CREATOR: gd-jpeg v1.0 (using IJG JPEG v62), quality = 60", baseline, precision 8, 800x533, frames 3                                                       
imageresizing.jpg: JPEG image data, JFIF standard 1.01, resolution (DPI), density 96x96, segment length 16, baseline, precision 8, 800x533, frames 3                                                                                                                            
orig.jpg:          JPEG image data, JFIF standard 1.01, resolution (DPI), density 300x300, segment length 16, baseline, precision 8, 3700x2467, frames 3                                                                                                                        
```

Since CImage default quality  is 60 for JPEG-images, try adding `&quality=90` and see if that changes anything.

**Comment by flobox** [comment]
Sadly no. We already played aorund with the quality operator. But this doesn't help anything regarding the color loss...

What I have found: http://stackoverflow.com/questions/8099252/resizing-images-with-gd-without-color-loss
http://php.net/manual/en/function.imagecreatetruecolor.php

**Comment by mosbth** [comment]
Okey. First. When we have this kind of issues "my image looks **** from what I expected" we need a way to deal with this.

1. Original image.
2. Image converted as expected results.
3. How to create the same image using cimage.

This you have.

Then I will use Imagemagick as another tool to create the same image.

So in this case I convert the original image using:

```
convert original.jpg -resize 800x imagemagick.jpg
```

Then we can play around to compare the images and use various options to create the same image using img.php options. 

A tool for that is in `webroot/compare/compare.php`.

It can look like this:

http://cimage.dbwebb.se/compare/compare.php?input1=https%3A%2F%2Fcloud.githubusercontent.com%2Fassets%2F3763232%2F10601524%2Fae9d7424-7712-11e5-8cf2-6873175854dd.jpg&input2=%2Fimg%2Fissue120%2Fimagemagick.jpg&input3=%2Fimg1.php%3Fsrc%3Dissue120%2Foriginal.jpg%26width%3D800&input4=%2Fimg1.php%3Fsrc%3Dissue120%2Foriginal.jpg%26width%3D800%26q%3D90&input5=%2Fimg1.php%3Fsrc%3Dissue120%2Foriginal.jpg%26width%3D800%26no-resample&input6=%2Fimg1.php%3Fsrc%3Dissue120%2Foriginal.jpg%26width%3D800%26sharpen&json=true&stack=true&bg=false

The link sets up the testcase and the testimages I prepared. Click it to review the testcase yourself.

So, now moving on to the issue.

The imagemagick version is more or less the same as imageresizing.net. Perhaps is imageresizing.net using imagemagick for resizing.

The img.php versions is loosing saturation. It does not depend on quality. It *might* depend on color profiles and how its handled in PHP GD. But I have not yet found a proper source for this.

Any ideas on why its like this, are most welcome.

**Comment by mosbth** [comment]
Ok. Its a matter of color space/color profile.

The original image has a colorspace of Adobe RGB (1998) which has a larger range of colors than the more common (for webusage) sRGB. PHP GD does resize the image and preserving the colors if the original image is stored with sRGB. PHP GD can reflect the colors in Adobe RGB (1998) which gives a *dull* impression where the image looses its sparkling colors.

I took the original image, opened it in Gimp which detected the colorspace and I answered yes to the question if I wanted to convert to sRGB colorspace. I then saved the image in both jpg and png format and resized both images using CImage and they both retained their vivid colors.

So, its basically a matter of how you deal with the source images. Use sRGB and it plays nicely. Use Adobe RGB (1998) and you get a loss in vivid colors.

Should CImage deal with this? Perhaps. 

There are several ways to detect and convert to sRGB colorspace. CImage could use external tools such as jpegicc or ImageMagic convert. CImage could also see if ImageMagick is installed as a PHP extension and use that to detect and convert to sRGB.

Or just ignore it. Food to thought. Hamburger. Lets think about it and see how big issue this is.

**Comment by flobox** [comment]
WOW...you are right... :)

I converted all our images into sRGB and it works! For me this solves the situation. Sure you are right, somehow it would be nice, if cimage could also deal with it. But again I am fine with this "fix".

Out for a burger!

**Comment by mosbth** [comment]
Okey, its to be included in v0.7.8. Just for the fun of it. You can check out the difference between the images here.

http://cimage.dbwebb.se/compare/compare.php?input1=%2Fimage%2Fhamburger.jpg&input2=%2Fimage%2Fhamburger.jpg%3Fso%26nc&input3=%2Fimage%2Fhamburger.jpg%3Fso%26nc%26srgb&input4=&input5=&input6=&json=false&stack=true&bg=false

The original image has a colorprofile of Adobe RGB (1998).

Its the option `&srgb` that does the magic by converting the image to sRGB and using the current color profile and changing color profile to a one working with GD. This option can be configured in the config file.

```
      /**
       * Convert the image to srgb before processing. Saves the converted
       * image in the sub cache dir. This option is default false but can
       * be changed to default true to do this conversion for all images.
       * This option requires PHP extension imagick and will silently fail
       * if that is not installed.
       *
       * Default value:
       *  srgb_dir:          $cachePath/srgb
       *  srgb_default:      false
       *  srgb_colorprofile: __DIR__ . '/../icc/sRGB_IEC61966-2-1_black_scaled.icc'
       */
       //'srgb_dir'  => __DIR__ . '/../cache/srgb/',
       //'srgb_default' => false,
       //'srgb_colorprofile' => __DIR__ . '/../icc/sRGB_IEC61966-2-1_black_scaled.icc',
```

Another visible example on this is the image car.png that has a colorprofile of "Color LCD". Check out this example to see the difference.

http://cimage.dbwebb.se/compare/compare.php?input1=%2Fimage%2Fcar.png&input2=%2Fimage%2Fcar.png%3Fso%26nc&input3=%2Fimage%2Fcar.png%3Fso%26nc%26srgb&input4=&input5=&input6=&json=false&stack=true&bg=false

One reason for including this is making it easier to test and debug various images. It could be useful to set the `srgb_default' => true` for testing but one should really convert all source images - that would be the most "proper" way to deal with this.
